crates/chia-bls/src/cached_bls.rs

```diff
@@ -0,0 +1,291 @@
+
+    pub fn generator(cache_size: Option<usize>) -> Self {
+        let cache: LruCache<[u8; 32], GTElement> =
+            LruCache::new(NonZeroUsize::new(cache_size.unwrap_or(50000)).unwrap());
+        Self { cache }
+    }
+
+    fn get_pairings<M: AsRef<[u8]>>(
+        &mut self,
+        pks: &[[u8; 48]],
+        msgs: &[M],
+        force_cache: bool,
+    ) -> Vec<GTElement> {
+        let mut pairings: Vec<Option<GTElement>> = vec![];
+        let mut missing_count: usize = 0;
+
+        for (pk, msg) in pks.iter().zip(msgs.iter()) {
+            let mut hasher = Sha256::new();
+            hasher.update(pk);
+            hasher.update(msg); // pk + msg
+            let h: [u8; 32] = hasher.finalize().into();
+
+            let pairing = self.cache.get(&h).cloned();
+
+            if !force_cache && pairing.is_some() {
+                // Heuristic to avoid more expensive sig validation with pairing
+                // cache when it's empty and cached pairings won't be useful later
+                // (e.g. while syncing)
+                missing_count += 1;
+                if missing_count > pks.len() / 2 {
+                    return vec![];
+                }
+            }
+
+            pairings.push(pairing);
+        }
+
+        // G1Element.from_bytes can be expensive due to subgroup check, so we avoid recomputing it with this cache
+        let mut pk_bytes_to_g1: HashMap<[u8; 48], PublicKey> = HashMap::new();
+        let mut ret: Vec<GTElement> = vec![];
+
+        for (i, pairing) in pairings.into_iter().enumerate() {
+            if let Some(pairing) = pairing {
+                // equivalent to `if pairing is not None`
+                ret.push(pairing);
+                continue;
+            }
+
+            let mut aug_msg = pks[i].to_vec();
+            aug_msg.extend_from_slice(msgs[i].as_ref()); // pk + msg
+            let aug_hash: Signature = hash_to_g2(&aug_msg);
+
+            let pk_parsed: &mut PublicKey = pk_bytes_to_g1
+                .entry(pks[i])
+                .or_insert_with(|| PublicKey::from_bytes(&pks[i]).unwrap());
+
+            let pairing: GTElement = aug_hash.pair(pk_parsed);
+            let mut hasher = Sha256::new();
+            hasher.update(&aug_msg);
+            let h: [u8; 32] = hasher.finalize().into();
+            self.cache.put(h, pairing.clone());
+            ret.push(pairing);
+        }
+
+        ret
+    }
+
+    pub fn aggregate_verify<M: AsRef<[u8]>>(
+        &mut self,
+        pks: &[[u8; 48]],
+        msgs: &[M],
+        sig: &Signature,
+        force_cache: bool,
+    ) -> bool {
+        let mut pairings: Vec<GTElement> = self.get_pairings(pks, msgs, force_cache);
+        if pairings.is_empty() {
+            let mut data = Vec::<(PublicKey, &[u8])>::new();
+            for (pk, msg) in pks.iter().zip(msgs.iter()) {
+                let pk = PublicKey::from_bytes_unchecked(pk).unwrap();
+                data.push((pk.clone(), msg.as_ref()));
+            }
```

Comment:
creating a new vector to push the key and message to is not necessary. You can pass in `pks` and `msgs` directly to `agg_ver()`